src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template

```diff
@@ -183,21 +184,56 @@ public class ScopedMemoryAccess {
     }
 
     @ForceInline
-    public void setMemory(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    public void setMemory(AbstractMemorySegmentImpl segment, byte value) {
         try {
-            setMemoryInternal(session, o, offset, bytes, value);
+            setMemoryInternal(segment, segment.byteSize(), value);
         } catch (ScopedAccessError ex) {
             throw ex.newRuntimeException();
         }
     }
 
+    // FILL_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    // Update the value for Aarch64 once 8338975 is fixed.
+    private static final long FILL_NATIVE_THRESHOLD = 1L << (Architecture.isAARCH64() ? 10 : 5);
+
     @ForceInline @Scoped
-    private void setMemoryInternal(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    private void setMemoryInternal(AbstractMemorySegmentImpl segment, long length, byte value) {
```

Comment:
All this logic should be in the memory segment impl class, and the calls to `UNSAFE` replaced with calls to `SCOPED_MEMORY_ACCESS`. This logic is complex for a scoped method, and you risk running into limitations of scoped methods (they cannot exceed 10 Java frames). In general, scoped methods are really meant to be wrappers around some JVM intrinsics.